skytap/client.go

```diff
@@ -287,22 +359,147 @@ func (c *Client) checkResponse(r *http.Response) error {
 		errorResponse.RequestID = strToPtr(requestID)
 	}
 
-	if code := r.StatusCode; code == http.StatusLocked ||
-		code == http.StatusTooManyRequests ||
-		code == http.StatusConflict ||
-		(code >= http.StatusInternalServerError && code <= 599) {
-		if retryAfter := r.Header.Get(headerRetryAfter); retryAfter != "" {
-			val, err := strconv.Atoi(retryAfter)
-			if err == nil {
-				errorResponse.RetryAfter = intToPtr(val)
-			} else {
-				errorResponse.RetryAfter = intToPtr(c.retryAfter)
-			}
+	if retryAfter := r.Header.Get(headerRetryAfter); retryAfter != "" {
+		val, err := strconv.Atoi(retryAfter)
+		if err == nil {
+			errorResponse.RetryAfter = intToPtr(val)
 		} else {
 			errorResponse.RetryAfter = intToPtr(c.retryAfter)
 		}
-		errorResponse.RequiresRetry = true
+	} else {
+		errorResponse.RetryAfter = intToPtr(c.retryAfter)
 	}
 
 	return errorResponse
 }
+
+func (c *Client) checkResourceStateUntilSatisfied(ctx context.Context, req *http.Request, state *environmentVMRunState) error {
+	runStateCheck := c.requiresChecking(state)
+	if runStateCheck > noRunStateCheck {
+		var ok bool
+		var err error
+		for i := 0; i < c.retryCount; i++ {
+			if runStateCheck == envRunStateCheck {
+				ok, err = c.getEnvironmentRunState(ctx, state.environmentID, state.environment)
+			} else {
+				ok, err = c.getVMRunState(ctx, state.environmentID, state.vmID, state.vm)
+			}
+			if err != nil {
```

Comment:
this can be simple if you break the loop in case `err!=nill` or `ok` and return err fewer branches